pkg/mei/mkhiclient_linux.go

```diff
@@ -0,0 +1,309 @@
+	"MEI_DEV_ID_ICH9_82X38": 0x29E4,
+	"MEI_DEV_ID_ICH9_3200":  0x29F4,
+	"MEI_DEV_ID_ICH9_6":     0x28B4,
+	"MEI_DEV_ID_ICH9_7":     0x28C4,
+	"MEI_DEV_ID_ICH9_8":     0x28D4,
+	"MEI_DEV_ID_ICH9_9":     0x28E4,
+	"MEI_DEV_ID_ICH9_10":    0x28F4,
+	"MEI_DEV_ID_ICH9M_1":    0x2A44,
+	"MEI_DEV_ID_ICH9M_2":    0x2A54,
+	"MEI_DEV_ID_ICH9M_3":    0x2A64,
+	"MEI_DEV_ID_ICH9M_4":    0x2A74,
+	"MEI_DEV_ID_ICH10_1":    0x2E04,
+	"MEI_DEV_ID_ICH10_2":    0x2E14,
+	"MEI_DEV_ID_ICH10_3":    0x2E24,
+	"MEI_DEV_ID_ICH10_4":    0x2E34,
+	"MEI_DEV_ID_IBXPK_1":    0x3B64,
+	"MEI_DEV_ID_IBXPK_2":    0x3B65,
+	"MEI_DEV_ID_CPT_1":      0x1C3A,
+	"MEI_DEV_ID_PBG_1":      0x1D3A,
+	"MEI_DEV_ID_PPT_1":      0x1E3A,
+	"MEI_DEV_ID_PPT_2":      0x1CBA,
+	"MEI_DEV_ID_PPT_3":      0x1DBA,
+	"MEI_DEV_ID_LPT_H":      0x8C3A,
+	"MEI_DEV_ID_LPT_W":      0x8D3A,
+	"MEI_DEV_ID_LPT_LP":     0x9C3A,
+	"MEI_DEV_ID_LPT_HR":     0x8CBA,
+	"MEI_DEV_ID_WPT_LP":     0x9CBA,
+	"MEI_DEV_ID_WPT_LP_2":   0x9CBB,
+	"MEI_DEV_ID_SPT":        0x9D3A,
+	"MEI_DEV_ID_SPT_2":      0x9D3B,
+	"MEI_DEV_ID_SPT_H":      0xA13A,
+	"MEI_DEV_ID_SPT_H_2":    0xA13B,
+	"MEI_DEV_ID_LBG":        0xA1BA,
+	"MEI_DEV_ID_BXT_M":      0x1A9A,
+	"MEI_DEV_ID_APL_I":      0x5A9A,
+	"MEI_DEV_ID_DNV_IE":     0x19E5,
+	"MEI_DEV_ID_GLK":        0x319A,
+	"MEI_DEV_ID_KBP":        0xA2BA,
+	"MEI_DEV_ID_KBP_2":      0xA2BB,
+	"MEI_DEV_ID_CNP_LP":     0x9DE0,
+	"MEI_DEV_ID_CNP_LP_4":   0x9DE4,
+	"MEI_DEV_ID_CNP_H":      0xA360,
+	"MEI_DEV_ID_CNP_H_4":    0xA364,
+	"MEI_DEV_ID_CMP_LP":     0x02e0,
+	"MEI_DEV_ID_CMP_LP_3":   0x02e4,
+	"MEI_DEV_ID_CMP_V":      0xA3BA,
+	"MEI_DEV_ID_CMP_H":      0x06e0,
+	"MEI_DEV_ID_CMP_H_3":    0x06e4,
+	"MEI_DEV_ID_CDF":        0x18D3,
+	"MEI_DEV_ID_ICP_LP":     0x34E0,
+	"MEI_DEV_ID_JSP_N":      0x4DE0,
+	"MEI_DEV_ID_TGP_LP":     0xA0E0,
+	"MEI_DEV_ID_MCC":        0x4B70,
+	"MEI_DEV_ID_MCC_4":      0x4B75,
+}
+
+// various ME-related constants
+const (
+	// ME - current working state set to normal
+	meHfs1CwsNormal = 0x5
+	// ME - current operation mode set to normal
+	meHfs1ComNormal = 0x0
+	// ME - CSE's firmware SKU is custom
+	meHfs3FwSkuCustom = 0x5
+
+	pciMEHfsts1 = 0x40
+	pciMEHfsts3 = 0x60
+)
+
+// MKHIClient is a client to send MKHI commands via MEI.
+type MKHIClient struct {
+	MEI *MEI
+}
+
+// MKHI command groups, see
+// https://github.com/coreboot/coreboot/blob/b8b8ec832360ada5a313f10938bb6cfc310a11eb/src/soc/intel/common/block/include/intelblocks/cse.h#L22
+const (
+	mkhiGroupIDCbm    = 0x0
+	mkhiGroupIDHMRFPO = 0x5
+	mkhiGroupIDGen    = 0xff
+)
+
+// MKHI HMRFPO command IDs, see
+// https://github.com/coreboot/coreboot/blob/b8b8ec832360ada5a313f10938bb6cfc310a11eb/src/soc/intel/common/block/include/intelblocks/cse.h#L33
+const (
+	mkhiHMRFPOEnable    = 0x1
+	mkhiHMRFPOGetStatus = 0x3
+)
+
+// pciAddressToInt converts the last part of a PCI address string to integer,
+// e.g. 0000:00:16.0 will return 0x16 .
+func pciAddressToInt(pciaddr string) (int, error) {
```

Comment:
I'm so sorry, one more... I don't see this func used anywhere, can it be deleted?

Reply:
whoops, sorry! It was used in a previous version of this code, just removed and committed again